underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsPositionedUnderFileInputStream.java

```diff
@@ -61,24 +71,51 @@ public int read(byte[] b) throws IOException {
     return read(b, 0, b.length);
   }
 
+  /**
+   * If there has been a number of sequential reads in a row,
+   * we move to regular buffered reads.
+   */
   @Override
   public int read(byte[] buffer, int offset, int length) throws IOException {
-    int bytesRead = ((FSDataInputStream) in).read(mPos, buffer, offset, length);
+    int bytesRead;
+    if (mSequentialReadCount >= SEQUENTIAL_READ_LIMIT) {
+      ((FSDataInputStream) in).seek(mPos);
+      mSequentialReadCount = 0;
+    }
+    if (mPos == ((FSDataInputStream) in).getPos()) {
+      // same position, use buffered reads as default
+      bytesRead = in.read(buffer, offset, length);
+    } else {
+      bytesRead = ((FSDataInputStream) in).read(mPos, buffer, offset, length);
+    }
     if (bytesRead > 0) {
       mPos += bytesRead;
     }
+    mSequentialReadCount++;
```

Comment:
Should this only be incremented if bytesRead > 0?

Reply:
done